_posts/2020-08-26-springboot-docker-image.md

```diff
@@ -0,0 +1,399 @@
+
+### Fat Jar Format
+If we inspect the jar produced after running the Maven build, we get this output:
+
+```
+META-INF/
+META-INF/MANIFEST.MF
+org/
+.
+org/springframework/boot/loader/
+.
+BOOT-INF/
+BOOT-INF/classes/
+.
+BOOT-INF/lib/
+BOOT-INF/classpath.idx
+```
+### Enabling the Layering Feature
+Release 2.3 supports layering by extracting parts of this fat jar in separate layers.
+
+The layering feature is explicitly enabled with the Spring Boot Maven plugin:
+
+```xml
+<plugin>
+  <groupId>org.springframework.boot</groupId>
+  <artifactId>spring-boot-maven-plugin</artifactId>
+  <configuration>
+    <layers>
+      <enabled>true</enabled>
+    </layers>
+  </configuration> 
+</plugin>
+```
+The contents of the resulting jar after building with Maven are shown: 
+```
+META-INF/
+.
+BOOT-INF/lib/
+.
+BOOT-INF/lib/spring-boot-jarmode-layertools-2.3.3.RELEASE.jar
+BOOT-INF/classpath.idx
+BOOT-INF/layers.idx
+```
+The output shows an additional jar named `spring-boot-jarmode-layertools` and a `layer.idx` file. The layering feature is provided by `spring-boot-jarmode-layertools` as explained in the next section.
+
+## Extracting the Dependencies in Separate Layers 
+
+We use a system property - `jarmode` set to value - `layertools` to launch the `spring-boot-jarmode-layertools` jar instead of the application. 
+
+Let us launch our jar with a `layertools jar mode` system property:
+```shell
+java -Djarmode=layertools -jar target/usersignup-0.0.1-SNAPSHOT.jar
+```
+Running this command produces the output containing available command options:
+```shell
+Usage:
+  java -Djarmode=layertools -jar usersignup-0.0.1-SNAPSHOT.jar
+
+Available commands:
+  list     List layers from the jar that can be extracted
+  extract  Extracts layers from the jar for image creation
+  help     Help about any command
+
+```
+The output shows the commands `list`, `extract`, and `help` with `help` being the default. Let us run the command with the `list` option :
+```shell
+java -Djarmode=layertools -jar target/usersignup-0.0.1-SNAPSHOT.jar list
+```
+```shell
+dependencies
+spring-boot-loader
+snapshot-dependencies
+application
+```
+We can see the list of dependencies that can be added as layers. 
+
+The default layers are:
+
+| Library Name        | Contents           | 
+| :------------------------ |:-----------------|
+| `dependencies`     | any dependency whose version does not contain SNAPSHOT |
+| `spring-boot-loader`      | jar loader classes      | 
+| `snapshot-dependencies` | any dependency whose version contains SNAPSHOT     |
+| `application` | application classes and resources      |
+
+The layers are defined in a `layers.idx` file in the order that they should be added to the Docker Image. These layers get cached in the host after the first pull since they do not change. Only the updated application layer is downloaded to the host which is faster because of the reduced size.
+
+## Building the Image with Dependencies Extracted in Separate Layers
```

Comment:
I don't understand the motivation for this and the next section. Why are we building a layered Docker image manually, when the Gradle and Maven plugins do that for us (as explained in the previous sections)? I'd rather remove this section and the next, to be honest.

Reply:
This is just an alternate way of containerizing. Either use buildpack or a docker. If we use docker, then we need to extract the layers in one stage and copy those to the final image. If we use buildpack, then this step is not required. I have edited the text stating this.